### DISCUSSION TITLE
Why use "RootEntityName" compare, not use "EntityName" in NHibernate/Engine /EntityKey.Equals0

### DISCUSSION THREAD

**Discussion by Lio98** [nhibernate/nhibernate-core#3478] [General] [answered]
NHibernate.Engine.EntityKey.cs
``` csharp
		public bool Equals(EntityKey other)
		{
			if (other == null)
			{
				return false;
			}

			return
				other.RootEntityName.Equals(RootEntityName)
				&& _persister.IdentifierType.IsEqual(other.Identifier, Identifier, _persister.Factory);
		}

```

**Comment by fredericDelaporte** [comment, marked as answer]
Because that is the right thing to do.

https://github.com/nhibernate/nhibernate-core/blob/07c367fb859086970737c35f24fe66c089a034e2/src/NHibernate/Persister/Entity/IEntityPersister.cs#L36-L40

Why should we use the `EntityName`? In case we have a key from a proxy of a base class, we would no more be able to infer that the key points the same entity than a key of the entity loaded from its concrete type.
